From c2c369f300d0c5ca0edf9a749c457252fe661992 Mon Sep 17 00:00:00 2001
From: Eduard Stefes <eduard.stefes@ibm.com>
Date: Fri, 8 Nov 2024 12:42:10 +0100
Subject: maint: fix s390 buffer flushes

Problem reported by Nick Rosbrook in:
https://bugs.launchpad.net/ubuntu/+source/rsyslog/+bug/2083700

align the behavior of dfltcc_inflate to do the same as gzip_inflate
when it hits a premature EOF
---
 dfltcc.c    | 7 ++++++-
 tests/hufts | 6 ++++--
 2 files changed, 10 insertions(+), 3 deletions(-)

Bug: https://lists.gnu.org/archive/html/bug-gzip/2024-12/msg00000.html
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/gzip/+bug/2083700
Last-Update: 2025-01-27

--- a/dfltcc.c
+++ b/dfltcc.c
@@ -21,6 +21,7 @@
 #include <stdalign.h>
 #include <stdbool.h>
 #include <stdlib.h>
+#include <errno.h>
 
 #ifdef HAVE_SYS_SDT_H
 # include <sys/sdt.h>
@@ -440,7 +441,11 @@
           if (fill_inbuf (1) == EOF)
             {
               /* Premature EOF.  */
-              return 2;
+              flush_outbuf();
+              errno = 0;
+              read_error();
+              /* Premature EOF.  */
+              __builtin_unreachable();
             }
           inptr = 0;
         }
--- a/tests/hufts
+++ b/tests/hufts
@@ -40,10 +40,12 @@
 
 printf '\037\213\010\000\060\060\060\060\060\060\144\000\000\000' > bug33501 \
   || framework_failure_
-printf '\ngzip: stdin: invalid compressed data--format violated\n' >exp33501 \
+printf '\ngzip: stdin: invalid compressed data--format violated\n' > exp33501.1 \
+  || framework_failure_
+printf '\ngzip: stdin: unexpected end of file\n' > exp33501.2 \
   || framework_failure_
 returns_ 1 gzip -d <bug33501 >out33501 2> err33501-raw || fail=1
 sed "$clean_stderr" err33501-raw > err33501 || framework_failure_
-compare exp33501 err33501 || fail=1
+compare exp33501.1 err33501 || compare exp33501.2 err33501 || fail=1
 
 Exit $fail
